plugins/inputs/t128_graphql/query_builder.go

```diff
@@ -1,20 +1,107 @@
 package t128_graphql
 
-import "strings"
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+
+	"github.com/Jeffail/gabs"
+)
 
 func buildQuery(entryPoint string, fields map[string]string, tags map[string]string) string {
-	var replacer = strings.NewReplacer("[", "(", "]", "\")", "/", "{", ":", ":\"")
-	query := "query {" + replacer.Replace(entryPoint) + "{"
+	query := "query "
+
+	var buf bytes.Buffer
+	w := io.Writer(&buf)
+	jsonObj := buildQueryObject(entryPoint, fields, tags)
+	traverseQueryObjPreOrder(jsonObj, w)
+
+	query += buf.String()
+	return query
+}
 
-	for _, element := range fields {
-		query += "\n" + element
+func traverseQueryObjPreOrder(jsonObj *gabs.Container, w io.Writer) {
+	jsonChildren, err := jsonObj.ChildrenMap()
+	if err != nil {
+		fmt.Println("error")
+		return
 	}
-	query = strings.TrimSpace(query)
-	for _, element := range tags {
-		query += "\n" + element
+
+	//store and sort the keys
+	keys := make([]string, len(jsonChildren))
+	i := 0
+	for k := range jsonChildren {
+		keys[i] = k
+		i++
 	}
+	sort.Strings(keys)
 
-	query += "\n" + strings.Repeat("}", strings.Count(query, "{"))
+	children := 0
+	for i, key := range keys {
+		if strings.HasPrefix(key, "$") {
```

Comment:
Maybe move `$` to a const to make it clear that is our predicate signification

Reply:
What is this used for? Are you thinking $ROUTER, $NODE etc? The monitoring agent substitutes those with the right "value" when creating the agent config. I dont think the actual telegraf input will see it unless its relying on the telegraf environment concept (which I'm not sure how we can leverage)

Reply:
The `$` indicates the item is a predicate. For example `name` in `allRouters[name:ComboEast]/node/...` is mapped to `$name`

Reply:
Can you give me a proper example - pardon my ignorance but I'm missing something here. How is that input you show converted to a predicate with $name.

Reply:
Yeah no problem sorry for not giving enough context. Lets so you have the following config:
- entrypoint = `"allRouters[name:ComboEast]"`
- fields = `map[string]string{"enabled": "enabled"}`
- tags = `map[string]string{"name": "name", "admin-status":  "state/adminStatus"}`

This `buildQueryObject` function will return a json object that looks like
`{"allRouters":{"$name":"ComboEast","enabled":"enabled","name":"name","state":{"adminStatus":"admin-status"}}}`

So that `[name:ComboEast]` -> `"$name":"ComboEast"`

Reply:
Is this the entrypoint for a deviceInterface?
```
entrypoint = "allRouter[name:$ROUTER]/nodes/nodes[name:$NODE]/nodes/deviceInterfaces/nodes"
```

Reply:
For example

```
query MyQuery {
  allRouters(name: "kaushik-home") {
    nodes {
      authority {
        routers {
          nodes {
            timeSeriesAnalytic(metric: SESSION_COUNT, transform: AVERAGE) {
              value
              timestamp
            }
          }
        }
      }
    }
  }
}
```